docs/design/build-system.md

```diff
@@ -0,0 +1,154 @@
+A given build will load these sources from disk, process them, and bundle them together.
+
+
+### TypeScript
+
+Angular builds rely heavily on TypeScript-specific functionality for Ahead-of-Time template compilation (AOT).
+Outside Angular CLI, this is performed by the Angular Compiler (`ngc`), provided by `@angular/compiler-cli`.
+To avail of Ahead-of-Time template compilation within a Webpack compilation we use and distribute the `@ngtools/webpack` Webpack plugin.
+
+Typescript sources are loaded from disk and compiled in-memory into Javascript files that are stored in a virtual file system and made available to Webpack.
+During compilation we also perform a number of code transformations using TypeScript transformers that enable automatic usage of AOT, internationalization features, and server-side rendering.
+
+AOT compilation requires loading HTML and CSS resources referenced on Angular Components. These are processed using a Webpack child compilation.
+
+The build system allows specifying replacements for specific files by replacing what path is loaded from the virtual file system.
+This is used for conditional loading of code at build time.
+
+### Stylesheets
+
+Two types of stylesheets are used in the build system: global stylesheets and component stylesheets.
+Global stylesheets are injected into the `index.html` file, while component stylesheets are loaded directly into compiled Angular components.
+
+The build system supports plain CSS stylesheets as well as the Sass, LESS and Stylus CSS pre-processors.
+Stylesheet processing functionality is provided by `sass-loader`, `less-loader`, `stylus-loader`, `postcss-loader`, `postcss-import`, augmented in the build system by custom webpack plugins.
+
+todo: style resources
+
+
+### Assets
+
+Assets in the build system refer specifically to a list of files or directories that are meant to be copied verbatim as build artifacts.
+These files are not processed and commonly include images, favicons, pdfs and other generic file types.
+They are loaded into the compilation using `copy-webpack-plugin`.
+
+
+### Scripts
+
+Scripts in the build system refer specifically to Javascript files that are meant to be loaded directly on `index.html` without being processed.
+They are loaded into the compilation using a custom webpack plugin.
+
+
+### Third party dependencies
+
+Third party dependencies are mostly inside `node_modules` and are referenced via imports in source files.
+Stylesheet third party dependencies are treated mostly the same as sources.
+
+JavaScript third party dependencies suffer a more involved process.
+They are first resolved to a folder in `node_modules` via [Node Module Resolution](https://nodejs.org/api/modules.html#modules_modules).
+Afterwards one of the entry points in that folder's `package.json` is used following the `es2015 > browser > module > main` priority list.
```

Comment:
I don't understand what this `es2015 > browser > module > main` priority list means? Is this the loader trying to choose the best form of the dependency to import?

Reply:
Not the loader, but rather webpack itself trying to resolve a given module to an entry point. Do this changes make it clearer?

> JavaScript third party dependencies suffer a more involved process.
They are first resolved to a folder in `node_modules` via [Node Module Resolution](https://nodejs.org/api/modules.html#modules_modules).
A given module might have several different entry points, for instance one for use in NodeJS and another one for using in the browser.
Each entry point is listed in under a name pointing at a js file in that module's `package.json`.
We use `es2015 > browser > module > main` a priority list, where the first key matched name determines which entry point to use.
For instance, for a module that has both `browser` and `main` entry points, we pick `browser`.

Reply:
That helps a bit but I'm still a little confused. Is this a standard in `package.json`? This sounds similar to the `main` property, but I'm not familiar with a standard which allows listing multiple entry points for different environments. Is this an NPM thing or an Angular thing?

> Each entry point is listed in under a name pointing at a js file in that module's package.json.

What "name" is this referring to in the `package.json`? Is it looking for literally `es2015.js`, `browser.js`, etc.? Or is it looking for some `browser` property?